ql/src/java/org/apache/hadoop/hive/ql/exec/FetchTask.java

```diff
@@ -78,8 +78,8 @@ public void initialize(QueryState queryState, QueryPlan queryPlan, TaskQueue tas
       if (source instanceof TableScanOperator) {
         TableScanOperator ts = (TableScanOperator) source;
         // push down projections
-        ColumnProjectionUtils.appendReadColumns(
-            job, ts.getNeededColumnIDs(), ts.getNeededColumns(), ts.getNeededNestedColumnPaths());
+        ColumnProjectionUtils.appendReadColumns(job, ts.getNeededColumnIDs(), ts.getNeededColumns(),
+                ts.getNeededNestedColumnPaths(), ts.conf.hasVirtualCols());
```

Comment:
nit: Shall we expose `hasVirtualCols` on TSOperator instead of exposing and using the `conf`?

Reply:
Unfortunately it is not consistent when we are expose or not. Example.:
https://github.com/apache/hive/blob/6626b5564ee206db5a656d2f611ed71f10a0ffc1/ql/src/java/org/apache/hadoop/hive/ql/exec/FetchTask.java#L86